# Data on used cars from eBay Kleinanzeigen
The aim of this project is to clean the sample data and analyze the included used car listings. This sample is comprised of 50,000 data points from a larger data set originally scraped from eBay Kleinanzeigen, a classifieds section of the German eBay website, and then uploaded to Kaggle. It can now be found [here](https://data.world/data-society/used-cars-data).

We will begin by reading in the csv file using Pandas, and taking a quick look at the data.

In [1]:
import pandas as pd
import numpy as np

autos = pd.read_csv('autos.csv', encoding='Latin-1')

In [2]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [3]:
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

## Cleaning the Data
At first glance, we can see that there are exactly 50,000 entries with twenty columns. The entries are comprised of two data types: "object" and "int64" (equivalent to Python's string/mixed and integer data types respectively). Five columns contain some null entries, which we might want to deal with. Many column labels are written in camelcase instead of the preferred snakecase. We will want to reformat the values under `prices` as integers while removing commas and dollar signs. Language might also be an issue, so it would probably be best to change German into English wherever possible.

We will begin by converting the columns to snakecase, and renaming some to improve descriptiveness.

In [4]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [5]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price_eur', 'ab_test',
                 'vehicle_type', 'registration_year', 'transmission', 'power_ps', 'model',
                 'odometer_km', 'registration_month', 'fuel_type', 'brand','unrepaired_damage',
                 'ad_created', 'n_of_pictures', 'postal_code', 'ad_last_seen']

In [6]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type price_eur  ab_test vehicle_type  registration_year  \
0  privat    Angebot    $5,000  control          bus               2004   
1  privat    Angebot    $8,500  control    limousine               1997   
2  privat    Angebot    $8,990     test    limousine               2009   
3  privat    Angebot    $4,350  control   kleinwagen               2007   
4  privat    Angebot    $1,350     test        kombi               2003   

  transmission  power_ps   model odometer_km  registration_month fuel_type  \
0      manuell       158  andere   150,000km                   3       lpg   
1    automatik       286     7er   150,000km                   6    benzin   
2      manuell       102    golf    70,000km                   7    benzin   
3    automatik        71  fortwo    70,000km                   6    benzin   
4      manuell         0   focus   150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  n_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00              0   
1         bmw              nein  2016-04-04 00:00:00              0   
2  volkswagen              nein  2016-03-26 00:00:00              0   
3       smart              nein  2016-03-12 00:00:00              0   
4        ford              nein  2016-04-01 00:00:00              0   

   postal_code         ad_last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

In addition to the conversion to snakecase, we chose to make several changes to the column names:
1. `price` to `price_eur`. The plan is to remove the dollar signs ("$") and commas from entries under this label in order to convert them into integers. After checking the website, *eBay Kleinanzeigen*, we confirmed that prices are in Euros.

2. `gearbox` to `transmission`. A gearbox is different from transmission in english, and the column entries were clearly transmission types.

3. `odometer` to `odometer_km`. Like the `price_eur` column, we will be converting these entries into integers.

4. `lastSeen` to `ad_last_seen`. The data entries are of advertisements of used cars, and "`lastSeen`" is not a clear enough description for the column.

In [7]:
autos.describe(include='all')

date_crawled         name  seller offer_type price_eur ab_test  \
count                 50000        50000   50000      50000     50000   50000   
unique                48213        38754       2          2      2357       2   
top     2016-03-10 15:36:24  Ford_Fiesta  privat    Angebot        $0    test   
freq                      3           78   49999      49999      1421   25756   
mean                    NaN          NaN     NaN        NaN       NaN     NaN   
std                     NaN          NaN     NaN        NaN       NaN     NaN   
min                     NaN          NaN     NaN        NaN       NaN     NaN   
25%                     NaN          NaN     NaN        NaN       NaN     NaN   
50%                     NaN          NaN     NaN        NaN       NaN     NaN   
75%                     NaN          NaN     NaN        NaN       NaN     NaN   
max                     NaN          NaN     NaN        NaN       NaN     NaN   

       vehicle_type  registration_year transmission      power_ps  model  \
count         44905       50000.000000        47320  50000.000000  47242   
unique            8                NaN            2           NaN    245   
top       limousine                NaN      manuell           NaN   golf   
freq          12859                NaN        36993           NaN   4024   
mean            NaN        2005.073280          NaN    116.355920    NaN   
std             NaN         105.712813          NaN    209.216627    NaN   
min             NaN        1000.000000          NaN      0.000000    NaN   
25%             NaN        1999.000000          NaN     70.000000    NaN   
50%             NaN        2003.000000          NaN    105.000000    NaN   
75%             NaN        2008.000000          NaN    150.000000    NaN   
max             NaN        9999.000000          NaN  17700.000000    NaN   

       odometer_km  registration_month fuel_type       brand  \
count        50000        50000.000000     45518       50000   
unique          13                 NaN         7          40   
top      150,000km                 NaN    benzin  volkswagen   
freq         32424                 NaN     30107       10687   
mean           NaN            5.723360       NaN         NaN   
std            NaN            3.711984       NaN         NaN   
min            NaN            0.000000       NaN         NaN   
25%            NaN            3.000000       NaN         NaN   
50%            NaN            6.000000       NaN         NaN   
75%            NaN            9.000000       NaN         NaN   
max            NaN           12.000000       NaN         NaN   

       unrepaired_damage           ad_created  n_of_pictures   postal_code  \
count              40171                50000        50000.0  50000.000000   
unique                 2                   76            NaN           NaN   
top                 nein  2016-04-03 00:00:00            NaN           NaN   
freq               35232                 1946            NaN           NaN   
mean                 NaN                  NaN            0.0  50813.627300   
std                  NaN                  NaN            0.0  25779.747957   
min                  NaN                  NaN            0.0   1067.000000   
25%                  NaN                  NaN            0.0  30451.000000   
50%                  NaN                  NaN            0.0  49577.000000   
75%                  NaN                  NaN            0.0  71540.000000   
max                  NaN                  NaN            0.0  99998.000000   

               ad_last_seen  
count                 50000  
unique                39481  
top     2016-04-07 06:17:27  
freq                      8  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

The columns labeled `seller`, and `offer_type` are cadidates for complete removal. They both contain 2 unique values, one of which exists in 1/50000 entries.

Several columns require more investigation, especially ones like `power_ps` and `n_of_pictures` which should be simple numbers but are returning null values.

`price_eur`, `registration_year`, `odometer_km`, etc. will be converted to integers to allow for calculations.

### Cleaning and Exploring `price_eur` and `odometer_km`
In order to convert the column entries into integers, we must first remove dollar signs($) and commas.

In [8]:
autos['price_eur'] = autos['price_eur'].str.replace('$','')
autos['price_eur'] = autos['price_eur'].str.replace(',','')
autos['price_eur'] = autos['price_eur'].astype(int)

autos['odometer_km'] = autos['odometer_km'].str.replace('km','')
autos['odometer_km'] = autos['odometer_km'].str.replace(',','')
autos['odometer_km'] = autos['odometer_km'].astype(int)

#### `price_eur'
We will now explore the same two columns, beginning with the price column, to look for any outliers in the data:

In [9]:
autos['price_eur'].unique().shape

(2357,)

In [10]:
autos['price_eur'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price_eur, dtype: float64

In [11]:
price_counts = autos['price_eur'].value_counts()
price_counts.sort_index()

0           1421
1            156
2              3
3              1
5              2
            ... 
10000000       1
11111111       2
12345678       3
27322222       1
99999999       1
Name: price_eur, Length: 2357, dtype: int64

Right away, we can see that there are both high and low priced outliers. From looking at current used car listings on eBay, it is clear that some numbers (low and high) are used as pre-negotiation placeholders, and are not the actual price of the car. Several low price entries are also for junk cars to be broken down and used for parts.

In order to facilitate our analysis we will be removing all listings below 500 euros, as they are not likely to be selling functioning cars. Anything above 350,000 euros will also be removed because there is a significant jump in cost for the next lisiting at exactly 999,990 euros.

In [12]:
autos = autos[autos['price_eur'].between(500,350000)]

In [13]:
autos['price_eur'].unique().shape

(2208,)

In [14]:
autos['price_eur'].describe()

count     45097.000000
mean       6320.659600
std        9261.841444
min         500.000000
25%        1500.000000
50%        3500.000000
75%        7900.000000
max      350000.000000
Name: price_eur, dtype: float64

In [15]:
price_counts = autos['price_eur'].value_counts()
price_counts.sort_index()

500       781
501         1
510         2
517         1
520         8
         ... 
265000      1
295000      1
299000      1
345000      1
350000      1
Name: price_eur, Length: 2208, dtype: int64

With the new parameters set, our data now contains 45,097 entries, nearly 5,000 fewer entries than before.

At first glance the new dataframe might appear to still have many outliers. 75% of all entries are priced at 7,900 euros or less, while the most expensive listing is 350,000 euros.

Upon deeper investigation (below), however, we can see that although there are clearly high-priced entries and potential outliers, these data points are not erroneous and provide important information.

For example, the most expensive listing is for a Porsche 991 costing 350,000 euros. A quick internet search in March of 2021 (5 years after this entry was listed) yielded several cars of the same make and model being sold for around 200,000 euros. Considering the age of our data, as well as the introduction of newer Porsche models, this Porsche 991 entry is likely not an erroneous outlier we can simply remove.

In [16]:
autos[autos['price_eur'] == 350000]

date_crawled         name  seller offer_type  price_eur  \
36818  2016-03-27 18:37:37  Porsche_991  privat    Angebot     350000   

       ab_test vehicle_type  registration_year transmission  power_ps model  \
36818  control        coupe               2016      manuell       500   911   

       odometer_km  registration_month fuel_type    brand unrepaired_damage  \
36818         5000                   3    benzin  porsche              nein   

                ad_created  n_of_pictures  postal_code         ad_last_seen  
36818  2016-03-27 00:00:00              0        70499  2016-03-27 18:37:37

#### `odometer_km`
Below, we will similarly analyze the "odometer_km" column:

In [17]:
autos['odometer_km'].unique().shape

(13,)

In [18]:
autos['odometer_km'].describe()

count     45097.000000
mean     125293.035013
std       39622.744927
min        5000.000000
25%      100000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [19]:
autos['odometer_km'].value_counts().sort_index()

5000        616
10000       237
20000       727
30000       765
40000       808
50000       996
60000      1131
70000      1189
80000      1385
90000      1676
100000     2031
125000     4838
150000    28698
Name: odometer_km, dtype: int64

Immediately, we encounter three problems with the `odometer_km` column:

1. There are only thirteen unique entries. These correspond with eBay Kleinanzeigen's search engine filters. Instead of providing the number of kilometers on each car, it tells us what range the entry falls under. For example, there are 808 entries which have the "odometer_km" value of "40000". This means that 808 vehicles have anywhere between 30,000 and 40,000 kilometers on them. This might not be especially useful because the ranges are not of the same magnitude.

2. The maximum mileage for the column is 150,000 kilometers. This heavily skews the data because over 75% of all entries fall within this range (125,000 - 150,000 km). 

3. There are many cars on the website which have over 150,000 kilometers on them, but there is no category that includes them. Perhaps this means there are no entries in our data that have over 150,000 kilometers. It could instead mean that everything above 150,000 was lumped into the same category. There is no way to know. Although based on how heavily skewed the data is, we cannot rule out the latter.

We are now left with two options. We could either:

* Convert entries in `odometer_km` to strings and add some information (e.g. "0-5000" instead of "5000"). This would limit our ability to do mathematical computations using the data, and would blindly assume that there are no vehicles in the data that have more than 150,000 km on them, despite the website having several today.
* Or we could remove the entries above "125000" in order to eliminate assumptions. This, however, would remove 28,698 entries.

In [20]:
autos = autos[autos['odometer_km'].between(0,125000)]

In [21]:
autos['odometer_km'].unique().shape

(12,)

In [22]:
autos['odometer_km'].describe()

count     16399.000000
mean      82056.222940
std       37144.143695
min        5000.000000
25%       50000.000000
50%       90000.000000
75%      125000.000000
max      125000.000000
Name: odometer_km, dtype: float64

In [23]:
autos['odometer_km'].value_counts().sort_index()

5000       616
10000      237
20000      727
30000      765
40000      808
50000      996
60000     1131
70000     1189
80000     1385
90000     1676
100000    2031
125000    4838
Name: odometer_km, dtype: int64

As shown above, we opted to remove the entries above "125000" from the `odometer_km` column, as it was the simpler option.

In [24]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 16399 entries, 2 to 49998
Data columns (total 20 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   date_crawled        16399 non-null  object
 1   name                16399 non-null  object
 2   seller              16399 non-null  object
 3   offer_type          16399 non-null  object
 4   price_eur           16399 non-null  int32 
 5   ab_test             16399 non-null  object
 6   vehicle_type        15339 non-null  object
 7   registration_year   16399 non-null  int64 
 8   transmission        15791 non-null  object
 9   power_ps            16399 non-null  int64 
 10  model               15700 non-null  object
 11  odometer_km         16399 non-null  int32 
 12  registration_month  16399 non-null  int64 
 13  fuel_type           15521 non-null  object
 14  brand               16399 non-null  object
 15  unrepaired_damage   14470 non-null  object
 16  ad_created          16

Here we noticed the indeces are no longer matching the length of our dataset, so it was quickly adjusted below.

In [25]:
autos = autos.set_axis(list(range(0,len(autos))))

In [26]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 16399 entries, 0 to 16398
Data columns (total 20 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   date_crawled        16399 non-null  object
 1   name                16399 non-null  object
 2   seller              16399 non-null  object
 3   offer_type          16399 non-null  object
 4   price_eur           16399 non-null  int32 
 5   ab_test             16399 non-null  object
 6   vehicle_type        15339 non-null  object
 7   registration_year   16399 non-null  int64 
 8   transmission        15791 non-null  object
 9   power_ps            16399 non-null  int64 
 10  model               15700 non-null  object
 11  odometer_km         16399 non-null  int32 
 12  registration_month  16399 non-null  int64 
 13  fuel_type           15521 non-null  object
 14  brand               16399 non-null  object
 15  unrepaired_damage   14470 non-null  object
 16  ad_created          16

In [27]:
autos.tail()

date_crawled                                              name  \
16394  2016-03-21 19:54:19        Mercedes_Benz_A_200__BlueEFFICIENCY__Urban   
16395  2016-03-10 19:37:38             Fiat_Grande_Punto_1.4_T_Jet_16V_Sport   
16396  2016-03-27 14:38:19  Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
16397  2016-04-02 14:44:48                    Fiat_500_C_1.2_Dualogic_Lounge   
16398  2016-03-08 19:25:42                Audi_A3_2.0_TDI_Sportback_Ambition   

       seller offer_type  price_eur  ab_test vehicle_type  registration_year  \
16394  privat    Angebot      17500     test    limousine               2012   
16395  privat    Angebot       4800  control   kleinwagen               2009   
16396  privat    Angebot      24900  control    limousine               2011   
16397  privat    Angebot      13200     test       cabrio               2014   
16398  privat    Angebot      22900  control        kombi               2013   

      transmission  power_ps     model  odometer_km  registration_month  \
16394      manuell       156  a_klasse        30000                  12   
16395      manuell       120    andere       125000                   9   
16396    automatik       239        q5       100000                   1   
16397    automatik        69       500         5000                  11   
16398      manuell       150        a3        40000                  11   

      fuel_type          brand unrepaired_damage           ad_created  \
16394    benzin  mercedes_benz              nein  2016-03-21 00:00:00   
16395       lpg           fiat              nein  2016-03-10 00:00:00   
16396    diesel           audi              nein  2016-03-27 00:00:00   
16397    benzin           fiat              nein  2016-04-02 00:00:00   
16398    diesel           audi              nein  2016-03-08 00:00:00   

       n_of_pictures  postal_code         ad_last_seen  
16394              0        58239  2016-04-06 22:46:57  
16395              0        68642  2016-03-13 01:44:51  
16396              0        82131  2016-04-01 13:47:40  
16397              0        73430  2016-04-04 11:47:27  
16398              0        35683  2016-04-05 16:45:07

### Exploring Dates
There are several columns whose values represent dates:

    date_crawled
    ad_created
    ad_last_seen
    registration_month
    registration_year

Entries in `date_crawled`, `ad_created`, and `ad_last_seen` are all strings. These will need to be changed so we can analyze them. The `registration_month` and `registration_year` columns are already represented as integers.

First we will examine the format of the entries in the three columns:

In [28]:
autos[['date_crawled','ad_created','ad_last_seen']].head()

date_crawled           ad_created         ad_last_seen
0  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
1  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
2  2016-03-31 19:48:22  2016-03-31 00:00:00  2016-04-06 14:17:52
3  2016-04-01 12:06:20  2016-04-01 00:00:00  2016-04-02 21:10:48
4  2016-03-28 20:50:54  2016-03-28 00:00:00  2016-04-01 06:45:30

In [29]:
autos[['date_crawled','ad_created','ad_last_seen']].tail()

date_crawled           ad_created         ad_last_seen
16394  2016-03-21 19:54:19  2016-03-21 00:00:00  2016-04-06 22:46:57
16395  2016-03-10 19:37:38  2016-03-10 00:00:00  2016-03-13 01:44:51
16396  2016-03-27 14:38:19  2016-03-27 00:00:00  2016-04-01 13:47:40
16397  2016-04-02 14:44:48  2016-04-02 00:00:00  2016-04-04 11:47:27
16398  2016-03-08 19:25:42  2016-03-08 00:00:00  2016-04-05 16:45:07

All entries appear to be written in the same format ("yyyy-mm-dd hh-mm-ss"). In order to look at the relative frequencies of the dates, we will only use at the first 10 characters of each entry (`.str[:10]`) and make sure to include non numerical entries (`dropna=False`). The relative frequencies will be shown as percentages instead of ratios.

In [30]:
(autos['date_crawled'].str[:10].value_counts(normalize=True, dropna=False)*100).sort_index()

2016-03-05    2.847735
2016-03-06    1.353741
2016-03-07    3.463626
2016-03-08    3.073358
2016-03-09    3.109946
2016-03-10    3.280688
2016-03-11    3.366059
2016-03-12    3.786816
2016-03-13    1.658638
2016-03-14    3.670956
2016-03-15    3.183121
2016-03-16    2.939204
2016-03-17    2.987987
2016-03-18    1.311056
2016-03-19    3.823404
2016-03-20    3.872187
2016-03-21    3.548997
2016-03-22    3.238002
2016-03-23    3.109946
2016-03-24    3.000183
2016-03-25    3.048966
2016-03-26    3.262394
2016-03-27    3.140435
2016-03-28    3.530703
2016-03-29    3.579487
2016-03-30    3.274590
2016-03-31    3.195317
2016-04-01    3.597780
2016-04-02    3.713641
2016-04-03    4.049027
2016-04-04    3.372157
2016-04-05    1.213489
2016-04-06    0.225624
2016-04-07    0.170742
Name: date_crawled, dtype: float64

In [31]:
(autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False)*100).sort_index()

2015-06-11    0.006098
2015-08-10    0.006098
2015-09-09    0.006098
2015-11-10    0.006098
2015-12-30    0.006098
                ...   
2016-04-03    4.061223
2016-04-04    3.347765
2016-04-05    1.152509
2016-04-06    0.243917
2016-04-07    0.152448
Name: ad_created, Length: 65, dtype: float64